src/storage/cookie.zig

```diff
@@ -0,0 +1,922 @@
+
+        var it = std.mem.splitScalar(u8, rest, ';');
+        while (it.next()) |attribute| {
+            const sep = std.mem.indexOfScalarPos(u8, attribute, 0, '=') orelse attribute.len;
+            const key_string = trim(attribute[0..sep]);
+
+            if (key_string.len > 8) {
+                // not valid, ignore
+                continue;
+            }
+
+            // Make sure no one changes our max length without also expanding the size of scrap
+            std.debug.assert(key_string.len <= 8);
+
+            const key = std.meta.stringToEnum(enum {
+                path,
+                domain,
+                secure,
+                @"max-age",
+                expires,
+                httponly,
+                samesite,
+            }, std.ascii.lowerString(&scrap, key_string)) orelse continue;
+
+            var value = if (sep == attribute.len) "" else trim(attribute[sep + 1 ..]);
+            switch (key) {
+                .path => {
+                    // path attribute value either begins with a '/' or we
+                    // ignore it and use the "default-path" algorithm
+                    if (value.len > 0 and value[0] == '/') {
+                        path = value;
+                    }
+                },
+                .domain => {
+                    if (value.len == 0) {
+                        continue;
+                    }
+                    if (value[0] == '.') {
+                        // leading dot is ignored
+                        value = value[1..];
+                    }
+
+                    if (std.mem.indexOfScalarPos(u8, value, 0, '.') == null) {
+                        // can't set a cookie for a TLD
+                        return error.InvalidDomain;
+                    }
+
+                    if (std.mem.endsWith(u8, host, value) == false) {
+                        return error.InvalidDomain;
+                    }
+                    domain = value;
+                },
+                .secure => secure = true,
+                .@"max-age" => max_age = std.fmt.parseInt(i64, value, 10) catch continue,
+                .expires => expires = DateTime.parse(value, .rfc822) catch continue,
+                .httponly => http_only = true,
+                .samesite => {
+                    same_site = std.meta.stringToEnum(Cookie.SameSite, std.ascii.lowerString(&scrap, value)) orelse continue;
+                },
+            }
+        }
+
+        if (same_site == .none and secure == null) {
+            return error.InsecureSameSite;
+        }
+
+        var arena = ArenaAllocator.init(allocator);
+        errdefer arena.deinit();
+        const aa = arena.allocator();
+        const owned_name = try aa.dupe(u8, cookie_name);
+        const owned_value = try aa.dupe(u8, cookie_value);
+        const owned_path = if (path) |p|
+            try aa.dupe(u8, p)
+        else
+            try defaultPath(aa, uri.path.percent_encoded);
```

Comment:
`defaultPath` don't always return an allocated value. is it a problem?

Reply:
Nope, the beauty of arena.

I hate that sometimes you feel like you need to dupe a string literal just to be able to safely call `free` on it when it isn't a string literal. Arena solves this.